src/EmbedIO/Security/IPBanningModule.cs

```diff
@@ -0,0 +1,301 @@
+            }
+            catch (Exception ex)
+            {
+                ex.Log(nameof(IPBanningModule), $"Invalid regex - '{pattern}'.");
+            }
+        }
+
+        internal void AddToWhitelist(IEnumerable<string> whitelist) =>
+            AddToWhitelistAsync(whitelist).GetAwaiter().GetResult();
+
+        internal async Task AddToWhitelistAsync(IEnumerable<string> whitelist)
+        {
+            if (whitelist?.Any() != true)
+                return;
+
+            foreach (var address in whitelist)
+            {
+                var addressees = await IPParser.Parse(address).ConfigureAwait(false);
+                _whitelist.AddRange(addressees.Where(x => !_whitelist.Contains(x)));
+            }
+        }
+        
+        /// <inheritdoc />
+        protected override Task OnRequestAsync(IHttpContext context)
+        {
+            ClientAddress = context.Request.RemoteEndPoint.Address;
+            if (!Blacklist.ContainsKey(ClientAddress)) 
+                return Task.CompletedTask;
+
+            context.SetHandled();
+            throw HttpException.Forbidden();
+        }
+        
+        /// <summary>
+        /// Releases unmanaged and - optionally - managed resources.
+        /// </summary>
+        /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_disposedValue) return;
+            if (disposing)
+            {
+                _whitelist.Clear();
+            }
+
+            _disposedValue = true;
+        }
+
+        private static void AddAccessAttempt(IPAddress address)
+        {
+            if (AccessAttempts.ContainsKey(address))
+                AccessAttempts[address].Add(DateTime.Now.Ticks);
+            else
+                AccessAttempts.TryAdd(address, new ConcurrentBag<long>() { DateTime.Now.Ticks });
+        }
+
+        private static void PurgeBlackList()
+        {
+            foreach (var k in Blacklist.Keys)
+            {
+                if (DateTime.Now.Ticks > Blacklist[k].BanUntil)
+                    Blacklist.TryRemove(k, out _);
+            }
+        }
+
+        private static void PurgeAccessAttempts()
+        {
+            var banDate = DateTime.Now.AddMinutes(-1).Ticks;
+            
+            foreach (var k in AccessAttempts.Keys)
+            {
+                var recentAttempts = new ConcurrentBag<long>(AccessAttempts[k].Where(x => x >= banDate));
+                if (!recentAttempts.Any())
+                    AccessAttempts.TryRemove(k, out _);
+                else
+                    Interlocked.Exchange(ref recentAttempts, AccessAttempts[k]);
+            }
+        }
+
+        private void UpdateBlackList()
+        {
+            var time = DateTime.Now.AddMinutes(-1).Ticks;
+            if ((AccessAttempts[ClientAddress]?.Where(x => x >= time).Count() >= _maxRetry))
```

Comment:
Beware that the indexer will not return `null` if the key is not found: it will throw a `KeyNotFoundException` instead. Once again, `TryGet` for the win.

Reply:
I already use `TryGet`